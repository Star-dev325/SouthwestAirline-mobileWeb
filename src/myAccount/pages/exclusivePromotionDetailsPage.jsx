// @flow
import React, { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';
import i18n from '@swa-ui/locale';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import withHideGlobalHeader from 'src/shared/enhancers/withHideGlobalHeader';
import Container from 'src/shared/components/container';
import PageHeader from 'src/shared/components/pageHeader';
import Button from 'src/shared/components/button';
import Icon from 'src/shared/components/icon';
import RegisteredMark from 'src/myAccount/components/registeredMark';
import {
  getExclusivePromotions,
  getPromotionDetails,
  registerUserExclusivePromotion
} from 'src/myAccount/actions/myAccountActions';
import { showDialog, hideDialog } from 'src/shared/actions/dialogActions';

import type { ExclusivePromotionSection } from 'src/myAccount/flow-typed/myAccount.types';
import type { Push } from 'src/shared/flow-typed/shared.types';

type Props = {
  push: Push,
  registerLink?: Link,
  isRegistered: boolean,
  promotionSections: Array<ExclusivePromotionSection>,
  showDialogFn: (*) => void,
  hideDialogFn: () => void,
  registerUserExclusivePromotionFn: (Link) => Promise<*>,
  isLoggedIn: boolean
};

type State = {
  promoIsRegistered: boolean
};

export class ExclusivePromotionDetailsPage extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      promoIsRegistered: props.isRegistered
    };
  }

  componentDidMount() {
    const { push, showDialogFn, hideDialogFn, promotionSections, isLoggedIn } = this.props;

    if (!isLoggedIn) {
      push('/login', null, { to: '/my-account/my-rapid-rewards/promotions' });
    } else {
      _.isEmpty(promotionSections) &&
        showDialogFn({
          active: true,
          name: 'promotion-id-can-not-match-wcm',
          message: i18n('SHARED__ERROR_MESSAGES__DEFAULT_API_ERROR_LATER'),
          buttons: [
            {
              label: 'OK',
              onClick: () => hideDialogFn()
            }
          ]
        });
    }
  }

  _onRegisterClick = (registerLink: Link) => {
    this.props.registerUserExclusivePromotionFn(registerLink).then(() => {
      this.setState({ promoIsRegistered: true });
    });
  };

  _renderPageHeader = () => (
    <PageHeader>
      <a
        onClick={() => this.props.push('/my-account/my-rapid-rewards/promotions')}
        ref="gobackIcon"
        className="goback-link"
      >
        <Icon type="keyboard-arrow-left" />
      </a>
      <span className="header">{i18n('MY_ACCOUNT__PROMOTION_DETAILS_PAGE__PROMOTION_DETAILS')}</span>
    </PageHeader>
  );

  _renderRegisterButton = (isRegistered: boolean, registerLink?: Link) =>
    (isRegistered ? (
      <RegisteredMark />
    ) : (
      <div className="wrapper">
        <Button color="grey" size="large" fluid onClick={() => registerLink && this._onRegisterClick(registerLink)}>
          {i18n('MY_ACCOUNT__PROMOTION_DETAILS_PAGE__PROMOTION_REGISTER')}
        </Button>
      </div>
    ));

  _renderWCMSections = (sections: Array<*>) =>
    !_.isEmpty(sections) && (
      <div className="promotion-details--content">
        {_.map(sections, (sectionItem, idx: number) => (
          <section key={idx}>
            <h4>{sectionItem.name}</h4>
            <div dangerouslySetInnerHTML={{ __html: sectionItem.content }} />
          </section>
        ))}
      </div>
    );

  render() {
    const { registerLink, promotionSections } = this.props;
    const { promoIsRegistered } = this.state;

    return (
      <div className="promotion-details">
        {this._renderPageHeader()}
        <div className="promotion-detail--item">
          <Container>
            {this._renderRegisterButton(promoIsRegistered, registerLink)}
            {promotionSections && this._renderWCMSections(promotionSections)}
          </Container>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  promotionSections: _.get(state, 'app.myAccountPages.exclusivePromotionDetailsPage.sections'),
  isRegistered: _.get(state, 'app.myAccountPages.exclusivePromotionDetailsPage.isRegistered'),
  registerLink: _.get(state, 'app.myAccountPages.exclusivePromotionDetailsPage._links.register'),
  isLoggedIn: _.get(state, 'app.account.isLoggedIn')
});

const mapDispatchToProps = {
  showDialogFn: showDialog,
  hideDialogFn: hideDialog,
  getExclusivePromotionsFn: getExclusivePromotions,
  getPromotionDetailsFn: getPromotionDetails,
  registerUserExclusivePromotionFn: registerUserExclusivePromotion
};

const enhancers = _.flowRight(
  withConnectedReactRouter,
  withHideGlobalHeader,
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(ExclusivePromotionDetailsPage);
