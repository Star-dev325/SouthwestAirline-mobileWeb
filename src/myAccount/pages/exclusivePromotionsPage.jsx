// @flow
import React, { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';
import i18n from '@swa-ui/locale';

import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import SubHeader from 'src/shared/components/subHeader';
import Container from 'src/shared/components/container';
import PromotionCard from 'src/myAccount/components/promotionCard';
import {
  getExclusivePromotions,
  getPromotionDetails,
  registerUserExclusivePromotion
} from 'src/myAccount/actions/myAccountActions';

import type {
  ExclusivePromotionPageType,
  ExclusivePromotionType,
  ExclusivePromotionDetailsResponseType
} from 'src/myAccount/flow-typed/myAccount.types';
import type { Push } from 'src/shared/flow-typed/shared.types';

type Props = {
  push: Push,
  exclusivePromotions: ExclusivePromotionPageType,
  getExclusivePromotionsFn: () => void,
  getPromotionDetailsFn: (Link) => Promise<ExclusivePromotionDetailsResponseType>,
  registerUserExclusivePromotionFn: (Link) => void
};

export class ExclusivePromotionsPage extends Component<Props> {
  componentDidMount() {
    const { getExclusivePromotionsFn } = this.props;

    getExclusivePromotionsFn();
  }

  _onPromotionDetailClick = (detailsLink: Link) => {
    const { push, getPromotionDetailsFn } = this.props;

    getPromotionDetailsFn(detailsLink).then(() => {
      push('/my-account/promotion-detail');
    });
  };

  _onRegisterClick = (registerLink: Link) => {
    const { registerUserExclusivePromotionFn } = this.props;

    registerUserExclusivePromotionFn(registerLink);
  };

  _renderPromotionGroup = (filteredPromotionList: Array<ExclusivePromotionType>, isRegistered: boolean = false) => (
    <div className="promotions-page--group">
      <label>
        {filteredPromotionList.length}{' '}
        {isRegistered
          ? i18n('MY_ACCOUNT__PROMOTIONS_PAGE__REGISTERED_PROMOTIONS')
          : i18n('MY_ACCOUNT__PROMOTIONS_PAGE__ELIGIBLE_PROMOTIONS')}
      </label>
      {_.map(filteredPromotionList, (promotionItem, index: number) => (
        <PromotionCard
          key={`exclusive-promotion-${index}`}
          promotion={promotionItem}
          onPromotionDetailClick={this._onPromotionDetailClick}
          onRegisterClick={this._onRegisterClick}
        />
      ))}
    </div>
  );

  render() {
    const { exclusivePromotions } = this.props;

    if (_.isEmpty(exclusivePromotions)) return null;

    const { numberOfEligiblePromotions, eligiblePromotions, numberOfRegisteredPromotions, registeredPromotions } =
      exclusivePromotions;
    const numberOfPromotions = numberOfEligiblePromotions + numberOfRegisteredPromotions;

    if (numberOfPromotions === 0) {
      return (
        <Container className="promotions-page">
          <div className="promotions-page--group">
            <label>{i18n('MY_ACCOUNT__PROMOTIONS_PAGE__NO_PROMOTIONS_AVAILABLE')}</label>
          </div>
        </Container>
      );
    }

    return (
      <div className="my-rapid-rewards">
        <SubHeader title={'My Rapid Rewards'} />
        <Container className="promotions-page">
          {!_.isEmpty(eligiblePromotions) && this._renderPromotionGroup(eligiblePromotions)}
          {!_.isEmpty(registeredPromotions) && this._renderPromotionGroup(registeredPromotions, true)}
        </Container>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  exclusivePromotions: _.get(state, 'app.myAccountPages.exclusivePromotions')
});

const mapDispatchToProps = {
  getExclusivePromotionsFn: getExclusivePromotions,
  getPromotionDetailsFn: getPromotionDetails,
  registerUserExclusivePromotionFn: registerUserExclusivePromotion
};

const enhancers = _.flowRight(withConnectedReactRouter, connect(mapStateToProps, mapDispatchToProps));

export default enhancers(ExclusivePromotionsPage);
