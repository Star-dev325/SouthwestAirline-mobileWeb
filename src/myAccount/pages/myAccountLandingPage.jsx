// @flow
import i18n from '@swa-ui/locale';
import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getAccountInfoForLandingPage, setTripTypeForDetailsPage } from 'src/myAccount/actions/myAccountActions';
import AccountNameHeader from 'src/myAccount/components/accountNameHeader';
import MyAccountPanel from 'src/myAccount/components/myAccountPanel';
import MyPromoCodesPanel from 'src/myAccount/components/myPromoCodesPanel';
import MyTripsPanel from 'src/myAccount/components/myTripsPanel';
import RapidRewardsPanelEnrolled from 'src/myAccount/components/rapidRewardsPanelEnrolled';
import RapidRewardsPanelNotEnrolled from 'src/myAccount/components/rapidRewardsPanelNotEnrolled';
import UnusedFundsPanel from 'src/myAccount/components/unusedFundsPanel';
import { CAR, FLIGHT } from 'src/myAccount/constants/upcomingTripType';
import {
  getCarRetrieveReservationInfoFromTrip,
  getRetrieveReservationInfoFromTrip
} from 'src/myAccount/helpers/upcomingTripsHelper';
import Container from 'src/shared/components/container';
import SubHeader from 'src/shared/components/subHeader';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import { buildPathWithParamAndQuery } from 'src/shared/helpers/pathUtils';
import { getNormalizedRoute } from 'src/shared/helpers/urlHelper';
import { clearFlightReservation, retrieveCarReservation } from 'src/viewReservation/actions/viewReservationActions';
import DynamicPlacement from 'src/wcm/components/dynamicPlacement';

import type { DynamicPlacementResponse } from 'src/wcm/flow-typed/wcm.types';
import type {
  ExclusivePromotionPageType,
  UnusedFundsContentType,
  PromoCodesContentType
} from 'src/myAccount/flow-typed/myAccount.types';
import type {
  CustomerInfoType,
  Push,
  RapidRewardsDetailsType,
  UpcomingTripPage
} from 'src/shared/flow-typed/shared.types';
import type {
  CarReservationType,
  RetrieveReservationRequestType
} from 'src/viewReservation/flow-typed/viewReservation.types';

type Props = {
  accountPagePlacements?: {
    banner01: DynamicPlacementResponse,
    promoCodeContentModule?: PromoCodesContentType,
    unusedFundsContentModule: ?UnusedFundsContentType
  },
  clearFlightReservationFn: () => void,
  customerInfo: CustomerInfoType,
  exclusivePromotions: ?ExclusivePromotionPageType,
  getAccountInfoForLandingPageFn: () => void,
  IsExclusivePromotionsHidden: boolean,
  isLoggedIn: boolean,
  isTierStatusPending: boolean,
  PROMO_CODE_IN_MY_ACCOUNT: boolean,
  push: Push,
  rapidRewardsDetails: ?RapidRewardsDetailsType,
  retrieveCarReservationFn: (retrieveReservationRequest: RetrieveReservationRequestType) => Promise<CarReservationType>,
  setTripTypeForDetailsPageFn: (string) => void,
  upcomingTrips: Array<UpcomingTripPage>,
  UNUSED_FUNDS: boolean
};

export class MyAccountLandingPage extends Component<Props> {
  componentDidMount() {
    const { getAccountInfoForLandingPageFn, isLoggedIn, push } = this.props;

    isLoggedIn ? getAccountInfoForLandingPageFn() : push('/login', null, { to: '/my-account' });
  }

  _getAccountNumberHeaderPros = () => {
    const { customerInfo, rapidRewardsDetails } = this.props;

    return {
      fullName: `${_.capitalize(customerInfo.name.firstName)} ${_.capitalize(customerInfo.name.lastName)}`,
      rapidRewardsNumber: rapidRewardsDetails?.isEnrolledInRapidRewards ? customerInfo.accountNumber : null
    };
  };

  _getEndOfYearInfo = () => {
    const { isTierStatusPending } = this.props;

    return {
      endOfYearFlag: isTierStatusPending,
      endOfYearContent: i18n('MY_ACCOUNT__PENDING_STATUS_TIER_TEXT')
    };
  };

  _gotoNextTrip = (nextTrip: UpcomingTripPage) => {
    const query = getRetrieveReservationInfoFromTrip(nextTrip);
    const name = _.pick(query, ['firstName', 'lastName']);
    const recordLocator = _.pick(query, 'recordLocator');

    this.props.push(
      buildPathWithParamAndQuery('/my-account/upcoming-trip-details/:tripIndex', { tripIndex: 0 }, recordLocator),
      null,
      null,
      name
    );
  };

  _onViewFundsBtnClick = () => {
    this.props.push(
      buildPathWithParamAndQuery(getNormalizedRoute({ routeName: 'travelFundsIndex' }), null, {
        clearFormData: false,
        clk: 'myaccounttravelfunds'
      })
    );
  };

  _onViewPromoCodesBtnClick = () => {
    this.props.push('/my-account/promo-codes');
  };

  _transitionToUpcomingTripDetailsPage = () => {
    const { clearFlightReservationFn, retrieveCarReservationFn, setTripTypeForDetailsPageFn, upcomingTrips } =
      this.props;
    const trip = _.first(upcomingTrips);

    setTripTypeForDetailsPageFn(trip.tripType);

    if (trip.tripType === FLIGHT) {
      clearFlightReservationFn();
      this._gotoNextTrip(trip);
    } else if (trip.tripType === CAR) {
      const request = getCarRetrieveReservationInfoFromTrip(trip);

      retrieveCarReservationFn(request).then(() => {
        this._gotoNextTrip(trip);
      });
    }
  };

  _renderContent = () => {
    const {
      accountPagePlacements,
      customerInfo,
      exclusivePromotions,
      IsExclusivePromotionsHidden,
      PROMO_CODE_IN_MY_ACCOUNT,
      push,
      rapidRewardsDetails,
      UNUSED_FUNDS,
      upcomingTrips
    } = this.props;

    if (!customerInfo || Object.keys(customerInfo).length === 0) return null;

    const { banner01, promoCodeContentModule, unusedFundsContentModule } = accountPagePlacements || {};
    const isEnrolledInRapidRewards = rapidRewardsDetails?.isEnrolledInRapidRewards || false;
    const promotionCount = exclusivePromotions
      ? _.add(exclusivePromotions.numberOfEligiblePromotions, exclusivePromotions.numberOfRegisteredPromotions)
      : 0;

    return (
      <Container ref="container">
        <AccountNameHeader {...this._getAccountNumberHeaderPros()} />
        {banner01 && (
          <MyAccountPanel>
            <DynamicPlacement {...banner01} data-qa="banner01" />
          </MyAccountPanel>
        )}
        {isEnrolledInRapidRewards ? (
          <RapidRewardsPanelEnrolled
            endOfYearInfo={this._getEndOfYearInfo()}
            IsExclusivePromotionsHidden={IsExclusivePromotionsHidden}
            onBenefitLinkClick={() => push('/my-account/tier-benefits-page')}
            onExclusivePromptClick={() => push('/my-account/my-rapid-rewards/promotions')}
            onRapidRewardsPanelClick={() => push('/my-account/my-rapid-rewards')}
            promotionsAvailable={promotionCount}
            rapidRewardsDetails={rapidRewardsDetails}
          />
        ) : (
          <RapidRewardsPanelNotEnrolled
            onEnrollClick={() => push('/my-account/enroll-rapid-rewards')}
            onGetDetailsClick={() => push('/about-rapid-rewards')}
          />
        )}
        <MyTripsPanel
          nextTrip={upcomingTrips[0]}
          numberOfUpcomingTrips={upcomingTrips.length}
          onClickBookATrip={() => push(getNormalizedRoute({ routeName: 'airBookingIndex' }))}
          onClickFindATrip={() => push('/view-reservation')}
          onClickPastFlights={() => push('/my-account/past-flights')}
          onClickNextTrip={this._transitionToUpcomingTripDetailsPage}
          onClickSavedFlights={() => push('/my-account/saved-flights')}
          onClickUpcomingTripCount={() => push('/my-account/upcoming-trips')}
          upcomingTripsApiResponseWasReceived={!_.isEmpty(upcomingTrips)}
        />
        {unusedFundsContentModule && (
          <UnusedFundsPanel
            onViewFundsBtnClick={this._onViewFundsBtnClick}
            UNUSED_FUNDS={UNUSED_FUNDS}
            unusedFundsContent={unusedFundsContentModule}
          />
        )}
        {PROMO_CODE_IN_MY_ACCOUNT && promoCodeContentModule && (
          <MyPromoCodesPanel
            onClick={this._onViewPromoCodesBtnClick}
            PROMO_CODE_IN_MY_ACCOUNT={PROMO_CODE_IN_MY_ACCOUNT}
            promoCodesContent={promoCodeContentModule}
          />
        )}
      </Container>
    );
  };

  render() {
    return (
      <div>
        <SubHeader title={i18n('MY_ACCOUNT__PAGE_HEADER_TITLE')} />
        {this._renderContent()}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  accountPagePlacements: _.get(state, 'app.myAccountPages.accountPagePlacements'),
  customerInfo: _.get(state, 'app.myAccountPages.customerAccountInfo.customerInfo'),
  exclusivePromotions: _.get(state, 'app.myAccountPages.exclusivePromotions', {}),
  IsExclusivePromotionsHidden: _.get(state, 'app.toggles.IsExclusivePromotionsHidden'),
  isLoggedIn: _.get(state, 'app.account.isLoggedIn'),
  isTierStatusPending: _.get(state, 'app.myAccountPages.customerAccountInfo.isTierStatusPending', false),
  PROMO_CODE_IN_MY_ACCOUNT: _.get(state, 'app.toggles.PROMO_CODE_IN_MY_ACCOUNT'),
  rapidRewardsDetails: _.get(state, 'app.myAccountPages.customerAccountInfo.rapidRewardsDetails'),
  upcomingTrips: _.get(state, 'app.upcomingTrips.upcomingTripsPage', []),
  UNUSED_FUNDS: _.get(state, 'app.toggles.UNUSED_FUNDS')
});

const mapDispatchToProps = {
  clearFlightReservationFn: clearFlightReservation,
  getAccountInfoForLandingPageFn: getAccountInfoForLandingPage,
  retrieveCarReservationFn: retrieveCarReservation,
  setTripTypeForDetailsPageFn: setTripTypeForDetailsPage
};

const enhancers = _.flowRight(withConnectedReactRouter, connect(mapStateToProps, mapDispatchToProps));

export default enhancers(MyAccountLandingPage);
