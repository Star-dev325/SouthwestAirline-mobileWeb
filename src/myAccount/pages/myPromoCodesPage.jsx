// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import i18n from '@swa-ui/locale';
import withBodyClass from 'src/shared/enhancers/withBodyClass';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import Container from 'src/shared/components/container';
import PageHeaderWithButtons from 'src/shared/components/pageHeaderWithButtons';
import PromoCodesList from 'src/shared/components/promoCodesList';
import { clearPromoCodes, getPromoCodes, getPromoCodesPagePlacements } from 'src/myAccount/actions/myAccountActions';
import DynamicPlacement from 'src/wcm/components/dynamicPlacement';
import { buildPathWithQuery } from 'src/shared/helpers/pathUtils';
import wcmTransitionTo from 'src/shared/helpers/wcmTransitionHelper';
import WcmLinkTypes from 'src/shared/constants/wcmLinkTypes';

import type { DynamicPlacementResponse } from 'src/wcm/flow-typed/wcm.types';
import type { PromoCodesPageType } from 'src/myAccount/flow-typed/myAccount.types';
import type { Push } from 'src/shared/flow-typed/shared.types';
import { DOLLAR, POINTS, REDEMPTION } from 'src/shared/constants/currencyTypes';
import { getNormalizedRoute } from 'src/shared/helpers/urlHelper';

type Props = {
  clearPromoCodesFn: () => void,
  promoCodesList: PromoCodesPageType,
  getPromoCodesFn: () => void,
  getPromoCodesPagePlacementsFn: () => void,
  placements?: { promoTop01?: DynamicPlacementResponse },
  isWebView: boolean,
  push: Push
};

export class MyPromoCodesPage extends Component<Props> {
  componentDidMount() {
    const { getPromoCodesFn, getPromoCodesPagePlacementsFn } = this.props;

    getPromoCodesFn();
    getPromoCodesPagePlacementsFn();
  }

  componentWillUnmount() {
    this.props.clearPromoCodesFn();
  }

  _onBookFlightClick = (promoCode: string, promoType: string) => {
    const currencyType = promoType === REDEMPTION ? POINTS : DOLLAR;
    const { push, isWebView } = this.props;

    if (isWebView) {
      const target = buildPathWithQuery('swaAppLink://airbooking', { isPoints: promoType === REDEMPTION, promoCode });

      wcmTransitionTo({ linkType: WcmLinkTypes.WEB_VIEW, target });
    } else {
      push(getNormalizedRoute({ routeName: 'airBookingIndex' }), null, { currencyType, promoCode });
    }
  };

  renderPromoCodesContent = () => {
    const { placements, isWebView, promoCodesList } = this.props;
    const { promoTop01 } = placements || {};

    return (
      <div className="placement-text-content">
        {promoTop01 && <DynamicPlacement {...promoTop01} data-qa="promoTop01" isWebView={isWebView} />}
        <PromoCodesList
          promoCodesList={promoCodesList}
          data-qa="promo-codes-list-content"
          onBookFlightClick={this._onBookFlightClick}
        />
      </div>
    );
  };

  renderNoPromoCodesMessage = () => (
    <Container>
      <p data-qa="no-promo-code-found-message" className="no-promo-code-found-message">
        {i18n('MY_ACCOUNT__NO_PROMO_CODE_FOUND_MESSAGE')}
      </p>
    </Container>
  );

  render() {
    const { isWebView, promoCodesList } = this.props;

    return (
      <div className="my-promo-codes">
        {!isWebView && <PageHeaderWithButtons showBackButton={true} title={i18n('MY_ACCOUNT__MY_PROMO_CODES')} />}
        {promoCodesList?.length ? this.renderPromoCodesContent() : this.renderNoPromoCodesMessage()}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  isWebView: _.get(state, 'app.webView.isWebView', false),
  placements: _.get(state, 'app.myAccountPages.promoCodesPagePlacements', {}),
  promoCodesList: _.get(state, 'app.myAccountPages.myAccountPromoCodes.promotions', [])
});

const mapDispatchToProps = {
  clearPromoCodesFn: clearPromoCodes,
  getPromoCodesFn: getPromoCodes,
  getPromoCodesPagePlacementsFn: getPromoCodesPagePlacements
};

const enhancers = _.flowRight(
  withBodyClass('bgpdkblue'),
  withConnectedReactRouter,
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(MyPromoCodesPage);
