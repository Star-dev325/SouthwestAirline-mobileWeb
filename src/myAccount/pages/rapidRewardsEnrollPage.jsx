// @flow
import React, { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';

import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import { enrollCustomerAccountForRR } from 'src/myAccount/actions/myAccountActions';
import { trackSubmitForm } from 'src/shared/analytics/actions/analyticsActions';
import { showDialog, hideDialog } from 'src/shared/actions/dialogActions';
import SubHeader from 'src/shared/components/subHeader';
import RapidRewardsEnrollForm from 'src/myAccount/components/rapidRewardsEnrollForm';
import i18n from '@swa-ui/locale';

import type { Replace } from 'src/shared/flow-typed/shared.types';
import type { FormData } from 'src/shared/form/flow-typed/form.types';
import type { RREnrollRequestDataType } from 'src/myAccount/flow-typed/myAccount.types';

type Props = {
  replace: Replace,
  isLoggedIn: boolean,
  minorAgeThreshold: number,
  dateOfBirth: string,
  enrollCustomerAccountForRRFn: (RREnrollRequestDataType) => Promise<*>,
  analyticsTrackSubmitFormFn: (string) => void,
  showDialogFn: (*) => Promise<*>,
  hideDialogFn: () => Promise<*>
};

export class RapidRewardsEnrollPage extends Component<Props> {
  _showCongratulationsPopup = () => {
    const { replace, showDialogFn, hideDialogFn } = this.props;

    showDialogFn({
      name: 'enroll-rapid-rewards-confirmation',
      title: i18n('MY_ACCOUNT__ENROLL_FOR_RAPID_REWARDS__CONFIRMATION_TITLE'),
      message: i18n('MY_ACCOUNT__ENROLL_FOR_RAPID_REWARDS__CONFIRMATION_MESSAGE'),
      buttons: [
        {
          label: 'OK',
          onClick: () =>
            hideDialogFn().then(() => {
              replace('/my-account');
            })
        }
      ]
    });
  };

  _onSubmit = ({ optInForEmailSubscriptions }: FormData) => {
    const { replace, isLoggedIn, analyticsTrackSubmitFormFn, enrollCustomerAccountForRRFn } = this.props;

    !isLoggedIn && replace('/my-account');
    analyticsTrackSubmitFormFn('my-account-enroll-for-rapid-rewards');

    enrollCustomerAccountForRRFn({ optInForEmailSubscriptions }).then(() => this._showCongratulationsPopup());
  };

  render() {
    const { dateOfBirth, minorAgeThreshold } = this.props;

    return (
      <div className="enroll-for-rapid-rewards">
        <SubHeader title={i18n('MY_ACCOUNT__ENROLL_IN_RAPID_REWARDS')} />
        <RapidRewardsEnrollForm
          formId={i18n('MY_ACCOUNT__RAPID_REWARDS_ENROLLMENT_FORM')}
          onSubmit={this._onSubmit}
          dateOfBirth={dateOfBirth}
          minorAgeThreshold={minorAgeThreshold}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  isLoggedIn: _.get(state, 'app.account.isLoggedIn'),
  minorAgeThreshold: _.get(state, 'app.wcmContent.applicationProperties.coppa.rrEnrollment.minAgeThreshold', 13),
  dateOfBirth: _.get(state, 'app.account.accountInfo.customerInfo.birthDate')
});

const mapDispatchToProps = {
  enrollCustomerAccountForRRFn: enrollCustomerAccountForRR,
  analyticsTrackSubmitFormFn: trackSubmitForm,
  showDialogFn: showDialog,
  hideDialogFn: hideDialog
};

const enhancers = _.flowRight(withConnectedReactRouter, connect(mapStateToProps, mapDispatchToProps));

export default enhancers(RapidRewardsEnrollPage);
