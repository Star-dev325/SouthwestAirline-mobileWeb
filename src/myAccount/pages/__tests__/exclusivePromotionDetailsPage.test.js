import i18n from '@swa-ui/locale';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { ExclusivePromotionDetailsPage } from 'src/myAccount/pages/exclusivePromotionDetailsPage';
import { createMockStoreWithRouterMiddleware } from 'test/unit/helpers/createMockStore';
import { createComponent } from 'test/unit/helpers/testingLibraryUtils';

describe('ExclusivePromotionDetailsPage', () => {
  let hideDialogFnMock;
  let pushMock;
  let registerUserExclusivePromotionFnMock;
  let showDialogFnMock;

  beforeEach(() => {
    hideDialogFnMock = jest.fn();
    pushMock = jest.fn();
    registerUserExclusivePromotionFnMock = jest.fn().mockResolvedValue('');
    showDialogFnMock = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('on mount', () => {
    it('if user is not logged in it push to login with a return pointed at promos list page', () => {
      createPageComponent({ isLoggedIn: false });

      expect(pushMock).toHaveBeenCalledWith('/login', null, { to: '/my-account/my-rapid-rewards/promotions' });
    });

    it('should play error popup if no WCM promo matches API response', () => {
      createPageComponent();

      expect(showDialogFnMock).toHaveBeenCalled();
      expect(showDialogFnMock.mock.calls[0][0]).toMatchObject({
        active: true,
        message: i18n('SHARED__ERROR_MESSAGES__DEFAULT_API_ERROR_LATER'),
        name: 'promotion-id-can-not-match-wcm'
      });
    });

    it('should render appropriate components if WCM and API IDs match', () => {
      const { container } = createPageComponent({
        promotionSections: [
          {
            content: 'First content block with information',
            name: 'First Header'
          },
          {
            content: 'Second content block with information',
            name: 'Second Header'
          }
        ]
      });

      expect(container).toMatchSnapshot();
    });

    it('should render register button if isRegistered is false', () => {
      const { container } = createPageComponent({ isRegistered: false });

      expect(container.querySelector('.button--fluid')).not.toBeNull();
      expect(container.querySelector('.registered-mark')).toBeNull();
    });

    it('should render register icon if isRegistered is true', () => {
      const { container } = createPageComponent({ isRegistered: true });

      expect(container.querySelector('.button--fluid')).toBeNull();
      expect(container.querySelector('.registered-mark')).not.toBeNull();
    });
  });

  describe('on click', () => {
    it('register button should register promotion then change UI to show RegisteredMark', async () => {
      const mockProps = {
        hideDialogFn: hideDialogFnMock,
        isLoggedIn: true,
        isRegistered: false,
        promotionSections: [
          {
            content: 'First content block with information',
            name: 'First Header'
          },
          {
            content: 'Second content block with information',
            name: 'Second Header'
          }
        ],
        push: pushMock,
        registerLink: {
          body: {
            promotionId: '1-23456'
          },
          href: '/v1/mobile-misc/feature/my-account/register-promotion',
          method: 'POST'
        },
        registerUserExclusivePromotionFn: registerUserExclusivePromotionFnMock,
        showDialogFn: showDialogFnMock
      };
      const state = {};
      const { container, rerender } = createComponent(ExclusivePromotionDetailsPage, { state, props: mockProps });
      const store = createMockStoreWithRouterMiddleware()({});

      expect(container.querySelector('.button--fluid')).not.toBeNull();
      expect(container.querySelector('.registered-mark')).toBeNull();

      fireEvent.click(container.querySelector('.button--fluid'));

      expect(registerUserExclusivePromotionFnMock).toBeCalledWith({
        href: '/v1/mobile-misc/feature/my-account/register-promotion',
        method: 'POST',
        body: {
          promotionId: '1-23456'
        }
      });

      await registerUserExclusivePromotionFnMock;

      rerender(
        <Provider store={store}>
          <ExclusivePromotionDetailsPage {...mockProps} />
        </Provider>
      );

      expect(screen.getByText('MY_ACCOUNT__REGISTERED_MARK__REGISTERED')).not.toBeNull();
    });

    it('back button should push user to exclusive promos list', () => {
      const { container } = createPageComponent({
        promotionSections: [
          {
            content: 'First content block with information',
            name: 'First Header'
          },
          {
            content: 'Second content block with information',
            name: 'Second Header'
          }
        ]
      });

      fireEvent.click(container.querySelector('.goback-link'));

      expect(pushMock).toBeCalledWith('/my-account/my-rapid-rewards/promotions');
    });
  });

  const createPageComponent = (props = {}) => {
    const defaultProps = {
      hideDialogFn: hideDialogFnMock,
      isLoggedIn: true,
      isRegistered: false,
      promotionSections: [],
      push: pushMock,
      registerLink: {
        body: {
          promotionId: '1-23456'
        },
        href: '/v1/mobile-misc/feature/my-account/register-promotion',
        method: 'POST'
      },
      registerUserExclusivePromotionFn: registerUserExclusivePromotionFnMock,
      showDialogFn: showDialogFnMock
    };

    const mergedProps = { ...defaultProps, ...props };
    const state = {};

    return createComponent(ExclusivePromotionDetailsPage, { state, props: mergedProps });
  };
});
