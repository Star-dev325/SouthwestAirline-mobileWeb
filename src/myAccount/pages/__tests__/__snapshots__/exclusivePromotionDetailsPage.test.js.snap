// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExclusivePromotionDetailsPage on mount should render appropriate components if WCM and API IDs match 1`] = `
<div>
  <div
    class="promotion-details"
  >
    <div>
      <div
        class="page-header"
        id=""
      >
        <a
          class="goback-link"
        >
          <i
            class="icon icon_keyboard-arrow-left"
          />
        </a>
        <span
          class="header"
        >
          MY_ACCOUNT__PROMOTION_DETAILS_PAGE__PROMOTION_DETAILS
        </span>
      </div>
      <div />
    </div>
    <div
      class="promotion-detail--item"
    >
      <div
        class="custom-container"
      >
        <div
          class="wrapper"
        >
          <button
            class="button button--fluid large button--grey"
            type="button"
          >
            MY_ACCOUNT__PROMOTION_DETAILS_PAGE__PROMOTION_REGISTER
          </button>
        </div>
        <div
          class="promotion-details--content"
        >
          <section>
            <h4>
              First Header
            </h4>
            <div>
              First content block with information
            </div>
          </section>
          <section>
            <h4>
              Second Header
            </h4>
            <div>
              Second content block with information
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</div>
`;
